pkg/storage/fingerprint_writer.go

```diff
@@ -0,0 +1,260 @@
+	}
+}
+
+type uintXorAggregate struct {
+	sum uint64
+}
+
+// add inserts one value into the running xor.
+func (a *uintXorAggregate) add(x uint64) {
+	a.sum = a.sum ^ x
+}
+
+// result returns the xor.
+func (a *uintXorAggregate) result() uint64 {
+	return a.sum
+}
+
+// Finish finalizes the underlying SSTWriter, and returns the aggregated
+// fingerprint for point keys.
+func (f *fingerprintWriter) Finish() (uint64, error) {
+	if err := f.sstWriter.Finish(); err != nil {
+		return 0, err
+	}
+	return f.xorAgg.result(), nil
+}
+
+// Close finishes and frees memory and other resources. Close is idempotent.
+func (f *fingerprintWriter) Close() {
+	if f.sstWriter == nil {
+		return
+	}
+	f.sstWriter.Close()
+	f.hasher.Reset()
+	f.xorAgg = nil
+	f.sstWriter = nil
+}
+
+var _ Writer = &fingerprintWriter{}
+
+// PutRawMVCCRangeKey implements the Writer interface.
+func (f *fingerprintWriter) PutRawMVCCRangeKey(key MVCCRangeKey, bytes []byte) error {
+	// We do not fingerprint range keys, instead, we write them to a Pebble SST.
+	// This is because range keys do not have a stable, discrete identity and so
+	// it is up to the caller to define a deterministic fingerprinting scheme
+	// across all returned range keys.ler to decide how to fingerprint them.
+	return f.sstWriter.PutRawMVCCRangeKey(key, bytes)
+}
+
+// PutRawMVCC implements the Writer interface.
+func (f *fingerprintWriter) PutRawMVCC(key MVCCKey, value []byte) error {
+	defer f.hasher.Reset()
+
+	// Hash the key/timestamp and value of the RawMVCC.
+	_, err := f.hasher.Write(key.Key)
+	if err != nil {
+		return errors.NewAssertionErrorWithWrappedErrf(err,
+			`"It never returns an error." -- https://golang.org/pkg/hash: %T`, f)
+	}
+	_, err = f.hasher.Write([]byte(key.Timestamp.String()))
+	if err != nil {
+		return errors.NewAssertionErrorWithWrappedErrf(err,
+			`"It never returns an error." -- https://golang.org/pkg/hash: %T`, f)
+	}
+	_, err = f.hasher.Write(value)
+	if err != nil {
+		return errors.NewAssertionErrorWithWrappedErrf(err,
+			`"It never returns an error." -- https://golang.org/pkg/hash: %T`, f)
+	}
+
+	f.xorAgg.add(f.hasher.Sum64())
+	return nil
+}
+
+// PutUnversioned implements the Writer interface.
+func (f *fingerprintWriter) PutUnversioned(key roachpb.Key, value []byte) error {
+	defer f.hasher.Reset()
+
+	// Hash the key and value in the absence of a timestamp.
+	_, err := f.hasher.Write(key)
+	if err != nil {
+		return errors.NewAssertionErrorWithWrappedErrf(err,
+			`"It never returns an error." -- https://golang.org/pkg/hash: %T`, f)
+	}
+	_, err = f.hasher.Write(value)
+	if err != nil {
+		return errors.NewAssertionErrorWithWrappedErrf(err,
+			`"It never returns an error." -- https://golang.org/pkg/hash: %T`, f)
+	}
+
+	f.xorAgg.add(f.hasher.Sum64())
+	return nil
+}
+
+// Unimplemented interface methods.
+
+// ApplyBatchRepr implements the Writer interface.
+func (f *fingerprintWriter) ApplyBatchRepr(repr []byte, sync bool) error {
+	panic("unimplemented")
```

Comment:
I'm fine either way, but in production I'm not sure the fingerprinting would ever be worth panic'ing the node for. Perhaps this could be an assertion error?

Also, given how much of this interface is unimplemented, it kinda makes me wonder if we want a smaller interface that we could use as the argument to MVCCExportToSST. Perhaps called `ExportWriter` or something.

Reply:
Good point, I like the idea of introducing a trimmed down `ExportWriter` interface. fwiw I don't think this would be a situation where we see an unexpected panic because these methods are called by us in `mvccExportToWriter` and we only use 3 of the many at the moment. It is possible someone forgets to test a change with the fingerprint option set on the ExportRequest so I am onboard with introducing a trimmed down interface.

Reply:
added an ExportWriter interface that is passed into `mvccExportToWriter` so that we can avoid all those stub methods in the fingerprint writer.